src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,126 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
+type Throttler struct {
+	sync.Mutex
+
+	keyState map[string]*keyContext
+	// Each entry in the queue should be unique to avoid unfair access
+	keyQueue []string
+
+	globalCurrentWeight int
+	globalMaxWeight     int
+}
+
+type keyContext struct {
+	// Requests for weight to be granted which are waiting.
+	waiting []request
+	// Currently granted weight for this key.
+	currentWeight int
+}
+
+type request struct {
+	blockCh chan struct{}
+	weight  int
+}
+
+// Acquire blocks until the requested weight is granted to the specified key.
+func (t *Throttler) Acquire(key string, weight int) {
```

Comment:
Drive by comment: past `context.Context` in here and have the channel wait below be a select that also waits on `ctx.Done()` so it properly respects cancellation. Would also need to have some indication that the wait returned due to cancelation not due to acquisition.

Reply:
Yeah agreed we definitely was to pass in `context` so that blocks to `Acquire` can be canceled